contracts/sol/MACI.sol

```diff
@@ -191,20 +239,138 @@ contract MACI is Ownable, DomainObjs {
+            uint8 x = 7 + i;
+            uint8 y = 7 + messageBatchSize + i * 2;
+            uint8 z = y + 1;
+            publicSignals[x] = _stateTreeRoots[i];
+            publicSignals[y] = _ecdhPubKeys[i].x;
+            publicSignals[z] = _ecdhPubKeys[i].y;
+        }
+
+        return publicSignals;
+    }
+
+    /*
+     * Update the postSignupStateRoot if the batch update state root proof is
+     * valid.
+     * @param _newStateRoot The new state root after all messages are processed
+     * @param _stateTreeRoots The intermediate state roots
+     * @param _ecdhPubKeys The public key used to generated the ECDH shared key
+     *                     to decrypt the message
+     * @param _proof The zk-SNARK proof
+     */
+    function batchProcessMessage(
+        uint256 _newStateRoot,
+        uint256[] memory _stateTreeRoots,
+        PubKey[] memory _ecdhPubKeys,
+        uint256[8] memory _proof
+    ) public {
+        
+        // Ensure that the array of state tree roots and the array of ECDH
+        // public keys are of the correct length
+        require(
+            _stateTreeRoots.length == messageBatchSize,
+            "MACI: incorrect _stateTreeRoots length"
+        );
+
+        require(
+            _ecdhPubKeys.length == messageBatchSize,
+            "MACI: incorrect _ecdhPubKeys length"
+        );
+
+        // Ensure that messageBatchStartIndex is within range
+        require(
+            messageBatchStartIndex <= messageTreeMaxLeafIndex - messageBatchSize,
+            "MACI: messageBatchStartIndex not within range"
+        );
+
+        // Assemble the public inputs to the snark
+        uint256[19] memory publicSignals = genBatchUstPublicSignals(
+            _newStateRoot,
+            _stateTreeRoots,
+            _ecdhPubKeys
+        );
+
+        // Ensure that each public input is within range of the snark scalar
+        // field.
+        // TODO: consider having more granular revert reasons
+        for (uint8 i=0; i < publicSignals.length; i++) {
+            require(
+                publicSignals[i] < SNARK_SCALAR_FIELD,
+                "MACI: each public signal must be lt the snark scalar field"
+            );
+        }
+
+        // Unpack the snark proof
+        (
+            uint256[2] memory a,
+            uint256[2][2] memory b,
+            uint256[2] memory c
+        ) = unpackProof(_proof);
+
+        // Verify the proof
+        bool isValid = batchUstVerifier.verifyProof(a, b, c, publicSignals);
+
+        require(isValid == true, "MACI: invalid batch UST proof");
+
+        // Increase the message batch start index to ensure that each message
+        // batch is processed in order
+        messageBatchStartIndex += messageBatchSize;
+
+        // The message batch start index should never exceed the maximum
+        // allowed value
+        assert(messageBatchStartIndex <= messageTreeMaxLeafIndex - messageBatchSize);
```

Comment:
I think its okay to allow the coordinator to process zero leaves at the end if they want. I think its more dangerous having this check here and not being able to process the last few votes.

Reply:
This check does not prevent the coordinator from processing zero-value leaves at the end. Instead, it prevents this function from being called more times than is acceptable.

E.g. if there is a message tree of maximum index 16, and the batch size is 4, this check ensures that `messageBatchStartIndex` never exceeds 12.